File: process_data.R
library(Matrix)

load_st_her2p <- function(st_id)
{
  data_dir_st <- "../Andersson_etal_2021"

  st_counts <- read.csv(sprintf("%s/ST-cnts/%s.tsv.gz", data_dir_st, st_id), 
                        sep = "\t", row.names = 1)
  st_counts <- as(t(as.matrix(st_counts)), "Matrix")
  
  st_meta <- read.csv(sprintf("%s/ST-spotfiles/%s_selection.tsv", data_dir_st, st_id), 
                      sep = "\t")
  rownames(st_meta) <- sprintf("%dx%d", st_meta$x, st_meta$y)
  
  pat_file <- sprintf("%s/ST-pat/lbl/%s_labeled_coordinates.tsv", data_dir_st, st_id)
  
  if(file.exists(pat_file))
  {
    pat <- read.csv(pat_file, sep = "\t")
    
    pat <- pat[which(!is.na(pat$x)), ]
    pat[, c("x2", "y2")] <- round(pat[, c("x", "y")])
    rownames(pat) <- sprintf("%dx%d", pat$x2, pat$y2)
    
    pat <- pat[rownames(st_meta), ]
    
    st_meta$pat_label <- pat$label
  }
  
  common_spots <- intersect(rownames(st_meta), colnames(st_counts))
  st_meta <- st_meta[common_spots, ]
  st_counts <- st_counts[, common_spots]
  
  return(list(counts = st_counts, meta = st_meta))
}

load_vis_tnbc <- function(vis_id)
{
  data_dir_st <- "../Wu_etal_2021_vis"
  
  vis_counts <- sprintf("%s/filtered_count_matrices/%s_filtered_count_matrix/", 
                        data_dir_st, vis_id)
  vis_counts <- Seurat::Read10X(vis_counts, gene.column = 1)
  
  vis_meta <- read.csv(sprintf("%s/metadata/%s_metadata_processed.csv", 
                               data_dir_st, vis_id), row.names = 1)
  
  vis_count <- vis_count[, rownames(vis_meta)]
  
  return(list(counts = vis_count, meta = vis_meta))
}

load_sc_wu <- function(subtype = "HER2+")
{
  data_dir_sc <- "../Wu_etal_2021_sc"
  
  sc_meta <- read.csv(sprintf("%s/metadata.csv", data_dir_sc), row.names = 1)
  
  if(subtype != 'all')
  {
    sc_meta <- sc_meta[which(sc_meta$subtype == subtype), ]
  }
  
  sc_count <- Seurat::Read10X(data_dir_sc, gene.column = 1)
  
  sc_count <- sc_count[, rownames(sc_meta)]
  
  return(list(counts = sc_count, meta = sc_meta))
}

load_libd_sample <- function(sample_name)
{
  libd_data_dir <- "../LIBD"
  
  layers <- read.csv(sprintf("%s/barcode_level_layer_map.tsv", libd_data_dir), sep = "\t", header = F)
  colnames(layers) <- c("spot_name", "sample_name", "layer")
  
  sample_counts <- Seurat::Read10X_h5(sprintf("%s/%s/filtered_feature_bc_matrix.h5", libd_data_dir, sample_name))
  
  sample_layers <- layers[which(layers$sample_name == sample_name), ]
  rownames(sample_layers) <- sample_layers$spot_name
  
  loc <- read.csv(sprintf("%s/%s/tissue_positions_list.txt", libd_data_dir, sample_name), 
                  header = F, row.names = 1)
  loc <- loc[which(loc$V2 == 1),  ]
  loc <- setNames(loc[, 2:3], c("Row", "Col"))
  
  spots <- intersect(rownames(loc), rownames(sample_layers))
  loc <- cbind(loc[spots, ], sample_layers[spots, "layer", drop = FALSE])
  loc$sample <- sample_name
  sample_counts <- sample_counts[, spots]
  rs <- rowSums(sample_counts)
  sample_counts <- sample_counts[which(rs > 0), ]
  
  colnames(sample_counts) <- rownames(loc) <- sprintf("%s_%s", sample_name, spots)
  
  return(list(counts = sample_counts, meta = loc))
}

load_libd_batch <- function(sample_names)
{
  dt <- c()
  for(sample_name in sample_names)
  {
    sample_data <- load_libd_sample(sample_name)
    
    if(is.null(dt))
    {
      dt <- sample_data
    }else
    {
      cg <- intersect(rownames(dt$counts), rownames(sample_data$counts))
      dt$counts <- cbind(dt$counts[cg, ], sample_data$counts[cg, ])
      dt$meta <- rbind(dt$meta, sample_data$meta)
    }
  }
  
  return(dt)
}

load_libd_experiment <- function(ref, srt)
{
  libd_data_dir <- "../LIBD"
  
  mtdata <- read.csv(sprintf("%s/libd_clinicaldata.csv", libd_data_dir))
  rownames(mtdata) <-  mtdata$SlideID <- as.character(mtdata$SlideID)
  
  sample_names <- mtdata$SlideID
  
  srt_samples <- c()
  if(srt %in% sample_names)
  {
    srt_samples <- srt
  }else
  {
    suppressWarnings(srt_samples <- sample_names[as.integer(srt)])
  }
  
  ref_samples <- c()
  if(ref == 'aggregated')
  {
    ref_samples <- sample_names
  }else if(ref == 'brain')
  {
    ref_samples <- sample_names[which(mtdata$Brain.Number == mtdata[srt_samples, "Brain.Number"])]
  }else if(ref == 'adjacent')
  {
    ref_samples <- sample_names[which(mtdata$Brain.Number == mtdata[srt_samples, "Brain.Number"] &
                                        mtdata$Position == mtdata[srt_samples, "Position"])]
  }else{
    stop("Invalid ref.")
  }
  
  ref_samples <- setdiff(ref_samples, srt_samples)
  return(list(ref = load_libd_batch(ref_samples), 
              srt = load_libd_batch(srt_samples)))
}

produce_mop_multicell <- function(sample_id = "mouse1_sample1", sc_data = NULL, multicell_meta = NULL, 
                                  seed = 1, diameter = 100, ann_column = "cluster_label")
{
  mop_dir <- "../MOp data"
  
  if(is.null(multicell_meta))
  {
    multicell_meta <- read.csv(sprintf("%s/multicell_meta.csv", mop_dir), 
                               colClasses = c("X" = "character"), row.names = 1)
  }
  
  sc_col <- sprintf("sc_seed_%d", seed)
  if(! (sc_col %in% colnames(multicell_meta)))
    stop("Seed not found!")
  
  d_col <- sprintf("spot_%g", diameter)
  if(! (d_col %in% colnames(multicell_meta)))
    stop("Diameter not found!")
  
  if(is.null(sc_data))
  {
    sc_data <- readRDS(sprintf("%s/scRNA_10X_v2_A/processed_data.rds", mop_dir))
  }
  
  mf_cells <- which(multicell_meta$sample_id == sample_id)
  spots <- multicell_meta[mf_cells, d_col]
  sc_cells <- multicell_meta[mf_cells, sc_col]
  
  srt_counts <- sc_data$counts[, sc_cells]
  srt_counts <- as.matrix(t(srt_counts))
  srt_counts <- t(rowsum(srt_counts, spots))
  srt_counts <- as(srt_counts, "Matrix")
  
  srt_composition <- as.data.frame.matrix(table(spots, sc_data$meta[sc_cells, ann_column]))
  
  srt_meta <- stringr::str_split(rownames(srt_composition), "X")
  srt_meta <- t(sapply(srt_meta, as.numeric))
  srt_meta <- setNames(as.data.frame(srt_meta), c("x", "y"))
  
  srt_id <- as.data.frame.matrix(table(spots, multicell_meta[mf_cells, "slice_id"]))
  srt_meta$slice_id <- colnames(srt_id)[apply(srt_id, 1, which.max)]
  
  remaining_sc <- setdiff(rownames(sc_data$meta), sc_cells)
  
  return(list(ref = list(counts = sc_data$counts[, remaining_sc], 
                         meta = sc_data$meta[remaining_sc, ]),
              srt = list(counts = srt_counts, 
                         composition = srt_composition,
                         meta = srt_meta)))
}

File: process_data.py
import pandas as pd
import scanpy as sc
import numpy as np
import scipy as sp
from scipy import sparse
import os

def load_libd_sample(sample_name):
    libd_dir = "../LIBD"
    
    layers = pd.read_csv(f"{libd_dir}/barcode_level_layer_map.tsv", sep="\t", header=None, dtype='str')
    layers.columns = ["spot_name", "sample_name", "layer"]
    
    layers = layers[layers["sample_name"] == sample_name].copy()
    layers.set_index("spot_name", inplace=True)
    
    loc = pd.read_csv(f"{libd_dir}/{sample_name}/tissue_positions_list.txt", sep=",", header=None, index_col=0)
    loc = loc[loc[1] == 1].iloc[:, 1:3].copy()
    loc.columns = ["Row", "Col"]
    
    spots = list(set(loc.index) & set(layers.index))
    
    loc = loc.loc[spots].copy()
    loc["layer"] = layers.loc[spots, "layer"]
    loc["sample"] = sample_name
    loc.index.name = "spot"
    
    adata = sc.read_10x_h5(f"{libd_dir}/{sample_name}/filtered_feature_bc_matrix.h5")
    adata.var_names_make_unique()
    adata = adata[spots, :].copy()
    
    rs = np.asarray(adata.X.sum(axis=0)).flatten()
    non_zero_indices = np.where(rs > 0)[0]
    
    adata = adata[:, non_zero_indices].copy()
    adata.obs = loc
    adata.obs_names = [f"{sample_name}_{spot}" for spot in adata.obs_names]

    return adata

def load_libd_batch(sample_names):
    return sc.concat([load_libd_sample(s) for s in sample_names], merge="same")

def load_libd_experiment(ref, srt):
    libd_dir = "../LIBD"
    
    mtdata = pd.read_csv(f"{libd_dir}/libd_clinicaldata.csv", dtype={"SlideID": str})
    mtdata.set_index("SlideID", inplace=True)
    
    sample_names = mtdata.index
    
    srt_samples = []
    if srt in sample_names:
        srt_samples.append(srt)
    else:
        try:
            srt_samples.append(sample_names[int(srt)])
        except (IndexError, ValueError):
            pass
    
    ref_samples = []
    if ref == 'aggregated':
        ref_samples = list(sample_names)
    elif ref == 'brain':
        brain_number = mtdata.loc[srt_samples[0], "Brain.Number"]
        ref_samples = sample_names[mtdata["Brain.Number"] == brain_number].tolist()
    elif ref == 'adjacent':
        brain_number = mtdata.loc[srt_samples[0], "Brain.Number"]
        position = mtdata.loc[srt_samples[0], "Position"]
        ref_samples = sample_names[(mtdata["Brain.Number"] == brain_number) & (mtdata["Position"] == position)].tolist()
    else:
        raise ValueError("Invalid ref.")
    
    ref_samples = set(ref_samples) - set(srt_samples)
    
    return load_libd_batch(ref_samples), load_libd_batch(srt_samples)

def sum_anndata_by(adata, col):
    adata.strings_to_categoricals()
    assert pd.api.types.is_categorical_dtype(adata.obs[col])

    cat = adata.obs[col].values
    indicator = sparse.coo_matrix(
        (
            np.broadcast_to(True, adata.n_obs),
            (cat.codes, np.arange(adata.n_obs))
        ),
        shape=(len(cat.categories), adata.n_obs),
    )

    return sc.AnnData(
        indicator @ adata.X,
        var=adata.var,
        obs=pd.DataFrame(index=cat.categories)
    )

def read_mop_sc():
    sc_dir = "../MOp data/scRNA_10X_v2_A"
    sc_data = sc.read_mtx(f"{sc_dir}/processed_data.mtx")
    sc_data = sc_data.T
    genes = pd.read_csv(f"{sc_dir}/processed_data_genes.csv")
    meta = pd.read_csv(f"{sc_dir}/processed_data_meta.csv", index_col = 0)
    sc_data.var_names = genes['x']
    sc_data.obs_names = meta.index
    sc_data.obs = meta

    return sc_data

def produce_mop_multicell(sample_id = "mouse1_sample1", sc_data = None,
                          seed=1, diameter=100, ann_column="cluster_label"):
    mop_dir = "../MOp data"
    multicell_meta = pd.read_csv(f"{mop_dir}/multicell_meta.csv", dtype={"X": str}, index_col=0)
    
    sc_col = f"sc_seed_{seed}"
    if sc_col not in multicell_meta.columns:
        raise ValueError("Seed not found!")
    
    d_col = f"spot_{diameter}"
    if d_col not in multicell_meta.columns:
        raise ValueError("Diameter not found!")

    if sc_data is None:
        sc_data = read_mop_sc()
    
    mf_cells = multicell_meta.index[multicell_meta['sample_id'] == sample_id].values
    spots = multicell_meta.loc[mf_cells, d_col].values
    sc_cells = multicell_meta.loc[mf_cells, sc_col].values
    
    srt_adata = sc_data[sc_cells, :].copy()
    srt_adata.obs['spot'] = spots
    srt_adata = sum_anndata_by(srt_adata, 'spot')
    
    srt_composition = pd.crosstab(index=spots, columns=sc_data.obs.loc[sc_cells, ann_column].values)
    srt_meta = pd.DataFrame(np.array(list(srt_composition.index.str.split("X", expand=True))))
    srt_meta.columns = ['x', 'y']
    srt_ids = pd.crosstab(index = spots, columns=multicell_meta.loc[mf_cells, 'slice_id'])
    srt_meta['slice_id'] = srt_ids.idxmax(axis=1)

    srt_meta.index = srt_adata.obs_names = srt_composition.index
    srt_adata.obs = srt_meta
    srt_adata.obsm['composition'] = srt_composition
    
    remaining_sc = list(set(sc_data.obs_names) -set(sc_cells))

    return sc_data[remaining_sc].copy(), srt_adata

def load_st_her2p(st_id):
    data_dir_st = "../Andersson_etal_2021"
    
    st_counts = pd.read_csv(f"{data_dir_st}/ST-cnts/{st_id}.tsv.gz", 
                            sep="\t", index_col=0)
    st_counts = sc.AnnData(st_counts)
    
    st_meta = pd.read_csv(f"{data_dir_st}/ST-spotfiles/{st_id}_selection.tsv", sep="\t")
    st_meta.index = [f"{x}x{y}" for x, y in zip(st_meta['x'], st_meta['y'])]
    
    pat_file = f"{data_dir_st}/ST-pat/lbl/{st_id}_labeled_coordinates.tsv"
    
    if os.path.exists(pat_file):
        pat = pd.read_csv(pat_file, sep="\t")
        pat = pat[~pat['x'].isna()]
        pat['x2'] = pat['x'].round().astype("int")
        pat['y2'] = pat['y'].round().astype("int")
        pat.index = [f"{x}x{y}" for x, y in zip(pat['x2'], pat['y2'])]
        pat = pat.loc[st_meta.index]
        
        st_meta['pat_label'] = pat['label'].values.tolist()
        
    common_spots = st_meta.index.intersection(st_counts.obs.index)
    st_meta = st_meta.loc[common_spots,:].copy()
    st_counts = st_counts[common_spots,:].copy()
    st_counts.obs = st_meta

    return st_counts

def read_mtx(dir):
    counts = sp.io.mmread(f"{dir}/single_file.mm").tocsr()
    genes = pd.read_csv(f"{dir}/single_file_genes.csv", index_col=0)
    barcodes = pd.read_csv(f"{dir}/single_file_barcodes.csv", index_col=0)
    adt = sc.AnnData(X = counts.T)
    adt.var_names = genes['x']
    adt.obs_names = barcodes['x']
    return adt

def load_vis_tnbc(vis_id):
    data_dir_vis = "../Wu_etal_2021_vis"
    
    vis_counts = read_mtx(f'{data_dir_vis}/filtered_count_matrices/{vis_id}_filtered_count_matrix/')
    vis_meta = pd.read_csv(f"{data_dir_vis}/metadata/{vis_id}_metadata_processed.csv", 
                           index_col=0)
    vis_counts = vis_counts[vis_meta.index].copy()
    vis_counts.obs = vis_meta
    
    return vis_counts

def load_sc_wu(subtype="HER2+"):
    sc_dir = "../Wu_etal_2021_sc"
    
    sc_meta = pd.read_csv(f"{sc_dir}/metadata.csv", index_col=0)
    if subtype != 'all':
        sc_meta = sc_meta[sc_meta['subtype'] == subtype].copy()
    
    sc_data = sc.read_mtx(f"{sc_dir}/matrix.mtx")
    sc_data = sc_data.T
    
    genes = pd.read_csv(f"{sc_dir}/genes.tsv", sep = "\t", header = None)
    barcodes = pd.read_csv(f"{sc_dir}/barcodes.tsv", sep = "\t", header = None)
    
    sc_data.var_names = genes[0].values
    sc_data.obs_names = barcodes[0].values
    
    sc_data = sc_data[sc_meta.index, :].copy()
    sc_data.obs = sc_meta
  
    return sc_data
 